### ISSUE TITLE
Assess FunASR

### ISSUE THREAD

**Issue by julien731** [nimblehq/audio-transcriber#89]
https://github.com/modelscope/FunASR

**Comment by julien731** [comment]
## FunASR assessment

Full write-up: `docs/assessments/funasr-assessment.md`

**Verdict: Monitor (primary); optional narrow pilot of SenseVoice-Small as an opt-in CPU fast-path. Do not adopt as a replacement.**

### Why

Our actual use case is European-language meetings (English/French — evidenced by the `fr` alignment entry and the English-only SER model). That is precisely FunASR's weakest area; its strength is Chinese/Asian languages, which we don't target. On accuracy, FunASR is at best lateral (if used only to wrap Whisper-large-v3-turbo, giving up its own selling point) and at worst a regression (its native models).

### Capability fit vs. recently-built features (#86/#87/#88)

These features are tightly coupled to WhisperX + PyAnnote internals, so a swap is a partial rewrite, not a drop-in:

- **Per-segment language detection across switches (#86/#87):** the value is our bespoke per-VAD-chunk, constrained-to-selected-set, confidence-thresholded loop with duration-weighted fallback (`multilingual_transcriber.py`). FunASR's language ID is a model output, not a tunable per-chunk loop — replicating current behavior means rebuilding that module.
- **Emotion on English (#88):** emotion2vec+ has a coarser label space than our 7→6 `RAW_LABEL_TO_CATEGORY` mapping; taxonomy and English gating would need re-deriving and re-validating, with user-visible category changes.
- **Interaction analysis (BR-3.1):** depends on raw PyAnnote turns *with overlaps* in `_diarize_and_assign`; cam++ would need equivalent overlapping-turn data or overlap/interruption detection regresses.
- **Word-level highlight (AC3):** SenseVoice timestamps may be too coarse.

### Architecture impact

Drop-in and partial (VAD/diarization-only) are both poor — they replace the cleanest parts of the pipeline and risk the as-built features. The only shape worth piloting is a **parallel, opt-in alternative engine** (mirroring the `WHISPER_*` env-config style), which fits the daemon-thread/job-queue model cleanly.

### Operational

- The **~17x-realtime-on-CPU claim** is the one genuinely interesting fact for GPU-less deploys (our CPU path is slow: float32, batch 4). Vendor-claimed — must be benchmarked.
- **Main integration hazard:** torch/torchaudio/**torchcodec** version conflict with our pins (`torch 2.8`, `torchcodec==0.7.0`) plus the PyTorch 2.6 `weights_only` monkeypatch.
- Model sizes (~234M–800M) and **MIT license** are non-issues.

### Uncertainty

No FunASR code was run. Performance multipliers, CPU viability, European-language accuracy, timestamp granularity, and emotion label mapping are all vendor-claimed or unverified. Verdict is deliberately conservative.

### Next steps (time-boxed)

1. **Dependency spike (½ day):** can `funasr` coexist with pinned torch/torchcodec on macOS/CPU? If not, parallel-backend is dead → pure Monitor.
2. **CPU benchmark (½ day):** SenseVoice-Small on real English/French clips — realtime factor + rough WER vs WhisperX.
3. Spot-check timestamp granularity and whether cam++ exposes overlapping turns.
4. **Decision gate:** only if (1) passes and (2) shows a clear speed win at acceptable accuracy, scope an "alternative engine" story behind config — multilingual/emotion/interaction explicitly out of scope for v1.
5. Otherwise close as Monitor — revisit if Asian-language meetings become a target or independent European-language benchmarks appear.

**Comment by julien731** [comment]
## Reassessment — Thai is the primary use case

Correcting the prior comment: it assumed European-language (EN/FR) meetings. A large share of meetings are in **Thai**, which changes the reasoning. The verdict direction is unchanged (do not adopt), but Thai actually *weakens* the FunASR case rather than strengthening it. Full write-up updated in `docs/assessments/funasr-assessment.md`.

**Verdict: Monitor. The only thing worth a narrow, GPU-based pilot is a Thai-WER head-to-head — Whisper-large-v3 (current) vs Fun-ASR-Nano on real Thai audio. Do not adopt as a replacement.**

### Why Thai weakens the FunASR case

1. **FunASR's compelling model doesn't speak Thai.** SenseVoice-Small — the ~170x-realtime-on-GPU, CPU-viable, emotion-bearing model that is FunASR's whole pitch — supports zh/en/ja/ko/yue **only, no Thai**. Paraformer-zh is zh/en only. So the headline speed and emotion advantages are irrelevant to our primary use case.
2. **The only Thai-capable FunASR model is Fun-ASR-Nano** — an 800M LLM/Qwen decoder that forfeits the CPU/speed pitch entirely (vendor points to vLLM/GPU for a mere 2-3x). For Thai, FunASR is a GPU-oriented choice, not a fast/cheap one. The Whisper-wrapper option is "just Whisper again," which we already run.
3. **We already do Thai end-to-end — the first pass missed this.** `transcriber.py:57` includes `"th"` in `ALIGNMENT_LANGUAGES`, and `:60-62` wires a dedicated Thai word-alignment model `CUSTOM_ALIGN_MODELS = {"th": "airesearch/wav2vec2-large-xlsr-53-th"}`, used by both the single-language and multilingual paths. Today Thai gets Whisper-large-v3 ASR + custom Thai wav2vec2 word timestamps + language-agnostic PyAnnote diarization, including in mixed-language meetings. No capability gap to fill.
4. **No Thai emotion win.** Emotion is English-only today (#88), and emotion2vec+ is largely EN/ZH-trained, not Thai-specific.

### The one thing that keeps this issue open

Thai is exactly where Whisper tends to be weaker, and Fun-ASR-Nano is an Asian-data-heavy LLM ASR — so it's genuinely plausible (not certain) it beats Whisper on **Thai WER**. That single benchmark is the only reason not to close #89 outright. Honest caveats: Fun-ASR-Nano's Thai WER is vendor-claimed/unverified and needs a GPU; and Thai's spaceless orthography makes word-level timestamps hard for both stacks, so even a WER win may not yield better word-level highlighting.

### Next steps (time-boxed)

1. **Thai-WER benchmark (decisive, ~1 day, GPU):** Fun-ASR-Nano vs current Whisper-large-v3 on 3-5 real Thai clips with references. If not clearly better, close #89.
2. **Dependency/GPU spike (gating):** can `funasr` + Fun-ASR-Nano coexist with pinned `torch==2.8` / `torchcodec==0.7.0`, and is a GPU runtime available for the target deploy?
3. **Thai timestamp + diarization-overlap spot-check** (player highlight on spaceless text; overlapping turns for interaction analysis BR-3.1).
4. **Gate:** only if (1) wins and (2) passes, scope a config-gated "alternative Thai engine" (GPU-required; advanced analysis features out of scope for v1).
5. Otherwise close as monitor — revisit on verified Thai benchmarks or a GPU deploy with a measured Thai-WER win.

**Comment by LauraGPT** [comment]
One narrow clarification that may make the proposed Thai spike cheaper and safer:

For Thai, I would not try to install FunASR directly into this app's existing WhisperX/PyAnnote environment first. Your `requirements.txt` already has the risky part called out: `torch`, `torchaudio`, and `torchcodec==0.7.0` are tied to the current stack. FunASR's vLLM path will typically bring its own matched `torch/torchaudio/torchvision` set, so the least invasive benchmark shape is a sidecar process/container exposing an HTTP API, while this app only calls it as an alternative ASR engine.

That turns your gating order into:

1. Run the Thai WER benchmark out-of-process first.
2. Only if it clearly wins, decide whether to keep it as a permanent sidecar or invest in dependency unification.
3. Keep diarization/interaction/word-highlight behavior owned by the current WhisperX/PyAnnote path until the ASR-only win is proven.

Also, SenseVoiceSmall is indeed not the Thai candidate. The Thai candidate is the 31-language Fun-ASR-Nano / MLT-Nano family; the FunASR docs list Thai among the 31 supported languages for the multilingual Nano variant. So your benchmark target should be the multilingual Nano path with `language=auto` or an explicitly pinned Thai setting, not SenseVoiceSmall.

Minimal smoke shape for the benchmark server side would be:

```bash
# separate env/container, not this repo's venv
pip install funasr vllm fastapi uvicorn python-multipart
python examples/industrial_data_pretraining/fun_asr_nano/serve_vllm.py \
  --model FunAudioLLM/Fun-ASR-Nano-2512 \
  --port 8899
```

Then the app-side experiment can stay a plain HTTP call, for example `POST /asr` or the OpenAI-compatible `POST /v1/audio/transcriptions`, depending on which FunASR server entrypoint you choose. That should answer the important question, Thai WER vs Whisper-large-v3 on your real clips, before spending time on a dependency spike inside the production venv.
